pyro/infer/autoguide/guides.py

```diff
@@ -656,44 +656,76 @@ def _loc_scale(self, *args, **kwargs):
         return self.loc, scale
 
 
-class AutoIAFNormal(AutoContinuous):
+class AutoNormalizingFlow(AutoContinuous):
     """
     This implementation of :class:`AutoContinuous` uses a Diagonal Normal
-    distribution transformed via a :class:`~pyro.distributions.transforms.AffineAutoregressive`
-    to construct a guide over the entire latent space. The guide does not depend on the model's
-    ``*args, **kwargs``.
+    distribution transformed via a sequence of bijective transforms
+    (e.g. various :mod:`~pyro.distributions.TransformModule` subclasses)
+    to construct a guide over the entire latent space. The guide does not
+    depend on the model's ``*args, **kwargs``.
 
     Usage::
 
-        guide = AutoIAFNormal(model, hidden_dim=latent_dim)
+        transform_init = partial(make_transforms, block_autoregressive,
+                                 repeats=2)
+        guide = AutoNormalizingFlow(model, transform_init)
         svi = SVI(model, guide, ...)
 
     :param callable model: a generative model
-    :param int hidden_dim: number of hidden dimensions in the IAF
+    :param init_transform_fn: a callable which when provided with the latent
+        dimension returns an instance of :class:`~torch.distributions.Transform`
+        , or :class:`~pyro.distributions.TransformModule` if the transform has
+        trainable params.
     :param callable init_loc_fn: A per-site initialization function.
         See :ref:`autoguide-initialization` section for available functions.
+
     """
 
-    def __init__(self, model, hidden_dim=None, init_loc_fn=init_to_median):
-        self.hidden_dim = hidden_dim
-        self.arn = None
+    def __init__(self, model, init_transform_fn, init_loc_fn=init_to_median):
+        self._init_transform_fn = init_transform_fn
+        self.transform = None
         super().__init__(model, init_loc_fn=init_loc_fn)
 
     def get_posterior(self, *args, **kwargs):
         """
         Returns a diagonal Normal posterior distribution transformed by
-        :class:`~pyro.distributions.transforms.iaf.InverseAutoregressiveFlow`.
-        """
-        if self.latent_dim == 1:
-            raise ValueError('latent dim = 1. Consider using AutoDiagonalNormal instead')
-        if self.hidden_dim is None:
-            self.hidden_dim = self.latent_dim
-        if self.arn is None:
-            self.arn = AutoRegressiveNN(self.latent_dim, [self.hidden_dim])
-
-        iaf = transforms.AffineAutoregressive(self.arn)
-        iaf_dist = dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.latent_dim]), [iaf])
-        return iaf_dist
+        a bijective transform.
+        """
+        if self.transform is None:
+            self.transform = self._init_transform_fn(self.latent_dim)
+        flow_dist = dist.TransformedDistribution(dist.Normal(0., 1.).expand([self.latent_dim]), self.transform)
+        return flow_dist
+
+
+class AutoIAFNormal(AutoNormalizingFlow):
```

Comment:
@martinjankowiak - I have only kept `AutoIAFNormal` here since we had it earlier, but if needed, we can add convenience wrappers for other transforms like `AutoBNAFNormal`. Right now it seems pretty convenient to just use `make_transform` and `AutoNormalizingFlow` directly, but I'm happy to add more if anyone requests it.

Reply:
yeah that seems fine for now. we'll probably want a `AutoBNAFNormal` but let's do that in another PR